spec/classes/init_spec.rb

```diff
@@ -7,45 +7,104 @@
     'foo.example.com'
   end
 
-  on_supported_os.each do |os, facts|
-    let :facts do
-      facts
+  on_supported_os.each do |os, os_facts|
+    let(:facts) do
+      os_facts
     end
-    context "on #{os}" do
-      context 'On an unknown OS with defaults for all parameters' do
-        it { is_expected.to raise_error(Puppet::Error) }
+
+    context 'on an unsupported OS' do
+      let(:facts) do
+        super().merge({
+                        'os' => {
+                          'family' => 'fakeLinux',
+                        }
+                      })
       end
 
-      context 'Test the default parameters (RedHat)' do
+      it { is_expected.to raise_error(Puppet::Error) }
+
+      context 'with fail_on_non_supported_os => false' do
+        let :params do
+          {
+            config_file_mode: '0755',
+            config_path: '/etc/cassandra',
+            fail_on_non_supported_os: false,
+            package_name: 'cassandra',
+            service_provider: 'base',
+            systemctl: '/bin/true'
+          }
+        end
+
         it do
+          expect(subject).to contain_file('/etc/cassandra/cassandra.yaml').with('mode' => '0755')
+          expect(subject).to contain_service('cassandra').with(provider: 'base')
+          expect(subject).to have_resource_count(6)
+        end
+      end
+    end
+
+    context "on #{os}" do
+      let(:facts) do
+        super().merge({
+                        cassandrarelease: '4.0.0',
+                      })
+      end
+
+      package_name = case os_facts[:os]['family']
+                     when 'RedHat' then 'cassandra22'
+                     else 'cassandra'
+                     end
+
+      systemctl_path = case os_facts[:os]['family']
+                       when 'RedHat' then '/usr/bin/systemctl'
+                       else '/bin/systemctl'
```

Comment:
AFAIK, `/bin` and `/usr/bin` are the same these days on most of the OSes (IIRC `/bin` is a symlink to `/usr/bin`). If so then just use `/usr/bin/systemctl` everywhere (except FreeBSD :-D)